packages/plugins/jwt/src/index.ts

```diff
@@ -82,8 +84,16 @@ export function useJWT(options: JwtPluginOptions): Plugin {
   return {
     async onRequestParse({ request, serverContext, url }) {
       const token = await getToken({ request, serverContext, url });
-      if (token != null) {
-        const signingKey = options.signingKey ?? (await fetchKey(jwksClient, token));
+      if (token) {
+        const decodedToken = decode(token, { complete: true });
+        if (!decodedToken?.header?.kid) {
```

Comment:
The "kid" header is mandatory only if the signing key is not already provided in the config :-) Because using a JKS is not the only way to provide it.

Tests are failing because most of them are actually using a statically defined key (for simplicity).

You should check if "signingKey" is defined, as it was before to solve this :-)

Reply:
Nice catch. I've added logic that should account for this (my setup uses public keys so I hadn't tested the situation where the `signingKey` is provided).

As an added bonus I circled back and fleshed out some logic where we can avoid having a stale key: if a verification fails, we force a refresh of the key from the endpoint, ensuring that we never throw a 401 as the result of a stale key. This should keep the cache as fresh as possible while still allowing a bunch of unnecessary requests to the JWKS endpoint, which will hopefully be the best of both worlds.

@EmrysMyrddin let me know what you think, and if I've missed any edge cases. :)

Reply:
@EmrysMyrddin, to avoid using mutable variables with the catch refresh mechanism, we could reach for something like this. This is nice because it abstracts much of the logic out into the helper function, but has the drawback of being probably slightly less legible code. Happy to opt for whichever approach is preferable to you:

```ts
import { createGraphQLError, Plugin } from 'graphql-yoga';
import jsonwebtoken, { Algorithm, JwtPayload } from 'jsonwebtoken';
import { JwksClient } from 'jwks-rsa';

const { decode } = jsonwebtoken;

export type JwtPluginOptions = JwtPluginOptionsWithJWKS | JwtPluginOptionsWithSigningKey;

export interface JwtPluginOptionsBase {
  /**
   * List of the algorithms used to verify the token
   *
   * Default: ["RS256"] (Allowed values: HS256, HS384, HS512, RS256, RS384, RS512, ES256, ES384, ES512, PS256, PS384, PS512, none)
   */
  algorithms?: Algorithm[];
  /**
   * The audience is the identifier of the API and is forwarded to your Auth API in order to specify for which API we are trying to authenticate our user for. E.g. if our API is hosted on `http://localhost:3000/graphql`, we would pass that value.
   */
  audience?: string;
  /**
   * For example: `https://myapp.auth0.com/`
   */
  issuer: string;
  /**
   * Once a user got successfully authenticated the authentication information is added on the context object under this field. In our resolvers we can then access the authentication information via `context._jwt.sub`.
   */
  extendContextField?: string;
  /**
   * Function to extract the token from the request object
   *
   * Default: Extracts the token from the Authorization header with the format `Bearer <token>`
   */
  getToken?: (params: {
    request: Request;
    serverContext: object | undefined;
    url: URL;
  }) => Promise<string | undefined> | string | undefined;
}

export interface JwtPluginOptionsWithJWKS extends JwtPluginOptionsBase {
  /**
   * The endpoint to fetch keys from.
   *
   * For example: https://example.com/.well-known/jwks.json
   */
  jwksUri: string;
  signingKey?: never;
}

export interface JwtPluginOptionsWithSigningKey extends JwtPluginOptionsBase {
  /**
   * Signing key to be used to verify the token
   * You can also use the jwks option to fetch the key from a JWKS endpoint
   */
  signingKey: string;
  jwksUri?: never;
}

export function useJWT(options: JwtPluginOptions): Plugin {
  if (!options.signingKey && !options.jwksUri) {
    throw new TypeError('You need to provide either a signingKey or a jwksUri');
  }

  if (options.signingKey && options.jwksUri) {
    throw new TypeError('You need to provide either a signingKey or a jwksUri, not both');
  }

  const { extendContextField = 'jwt', getToken = defaultGetToken } = options;

  const payloadByRequest = new WeakMap<Request, JwtPayload | string>();

  let jwksClient: JwksClient;
  const jwksCache: Map<string, string> = new Map();

  if (options.jwksUri) {
    jwksClient = new JwksClient({
      cache: true,
      rateLimit: true,
      jwksRequestsPerMinute: 5,
      jwksUri: options.jwksUri,
    });
  }

  return {
    async onRequestParse({ request, serverContext, url }) {
      const token = await getToken({ request, serverContext, url });
      if (token != null) {
        const verified = await fetchKeyAndVerify({ jwksClient, jwksCache, token, options });
        if (!verified) {
          throw unauthorizedError(`Unauthenticated`);
        }
        payloadByRequest.set(request, verified);
      }
    },
    onContextBuilding({ context, extendContext }) {
      if (context.request == null) {
        throw new Error(
          'Request is not available on context! Make sure you use this plugin with GraphQL Yoga.',
        );
      }
      const payload = payloadByRequest.get(context.request);
      if (payload != null) {
        extendContext({
          [extendContextField]: payload,
        });
      }
    },
  };
}

function unauthorizedError(message: string, options?: Parameters<typeof createGraphQLError>[1]) {
  return createGraphQLError(message, {
    extensions: {
      http: {
        status: 401,
      },
    },
    ...options,
  });
}

function verify(
  token: string,
  signingKey: string,
  options: Parameters<typeof jsonwebtoken.verify>[2],
): string | jsonwebtoken.JwtPayload | undefined {
  return new Promise((resolve, reject) => {
    jsonwebtoken.verify(
      token,
      signingKey,
      { ...options, algorithms: options?.algorithms ?? ['RS256'] },
      (err, result) => {
        if (err) {
          reject(unauthorizedError('Failed to decode authentication token. Verification failed.'));
        } else {
          resolve(result as JwtPayload);
        }
      },
    );
  });
}

interface FetchKeyAndVerifyOptions {
  jwksClient: JwksClient;
  jwksCache: Map<string, string>;
  token: string;
  options: JwtPluginOptions;
  shouldRefreshCache?: boolean;
}

async function fetchKeyAndVerify({
  jwksClient,
  jwksCache,
  token,
  options,
  shouldRefreshCache
}: FetchKeyAndVerifyOptions): Promise<JwtPayload | string | undefined> {
  const decodedToken = decode(token, { complete: true });
  const kid = decodedToken?.header?.kid;

  if (!kid) {
    throw unauthorizedError(`Failed to decode authentication token. Missing key id.`);
  }

  const signingKey = options?.signingKey ?? await (async () => {
    if (shouldRefreshCache) {
      jwksCache.delete(kid);
    }

    if (!jwksCache.has(kid)) {
      const secret = await jwksClient.getSigningKey(kid);
      const newSigningKey = secret?.getPublicKey();
      if (!newSigningKey) {
        throw unauthorizedError(`Failed to decode authentication token. Unknown key id.`);
      }
      jwksCache.set(kid, newSigningKey);
    }

    return jwksCache.get(kid)!;
  })();

  try {
    return verify(token, signingKey, options);
  } catch (error) {
    if (!shouldRefreshCache) {
      // If initial verification fails, attempt to refresh the key and retry verification
      return fetchKeyAndVerify({ jwksClient, jwksCache, token, options, shouldRefreshCache: true });
    }
    throw error;
  }
}

const defaultGetToken: NonNullable<JwtPluginOptions['getToken']> = ({ request }) => {
  const header = request.headers.get('authorization');
  if (!header) {
    return;
  }
  const [type, token] = header.split(' ');
  if (type !== 'Bearer') {
    throw unauthorizedError(`Unsupported token type provided: "${type}"`);
  }
  return token;
};
```

Reply:
@EmrysMyrddin it should be in a better spot now with this commit (https://github.com/dotansimha/graphql-yoga/pull/3161/commits/9cbe38ad24df2762c1c7a3a8fa71a5d1b0bcad04) to pass those tests, ran prettier and conformed all errors to the expected `Unauthorized` error text.